File: src/main/java/arcana/common/blocks/tiles/DefaultTile.java
package arcana.common.blocks.tiles;

import net.minecraft.block.BlockState;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.network.NetworkManager;
import net.minecraft.network.play.server.SUpdateTileEntityPacket;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.tileentity.TileEntityType;
import net.minecraftforge.common.util.Constants;

public class DefaultTile extends TileEntity {
    public DefaultTile(TileEntityType<?> tileEntityTypeIn) {
        super(tileEntityTypeIn);
    }

    @Override
    public SUpdateTileEntityPacket getUpdatePacket(){
        return new SUpdateTileEntityPacket(worldPosition, -1, save(new CompoundNBT()));
    }

    @Override
    public void onDataPacket(NetworkManager net, SUpdateTileEntityPacket packet){
        load(getBlockState(), packet.getTag());
    }

    @Override
    public CompoundNBT getUpdateTag() {
        return save(super.getUpdateTag());
    }

    @Override
    public void handleUpdateTag(BlockState state, CompoundNBT tag) {
        super.handleUpdateTag(state, tag);
        load(state, tag);
    }

    public void saveAndSync(){
        setChanged();
        level.setBlockAndUpdate(worldPosition, getBlockState());
    }
}

File: src/main/kotlin/arcana/common/blocks/tiles/DefaultTile.kt
package arcana.common.blocks.tiles

import net.minecraft.block.BlockState
import net.minecraft.nbt.CompoundNBT
import net.minecraft.network.NetworkManager
import net.minecraft.network.play.server.SUpdateTileEntityPacket
import net.minecraft.tileentity.TileEntity
import net.minecraft.tileentity.TileEntityType

open class DefaultTile(type: TileEntityType<*>) : TileEntity(type) {
    override fun getUpdatePacket(): SUpdateTileEntityPacket {
        return SUpdateTileEntityPacket(worldPosition, -1, save(CompoundNBT()))
    }

    override fun onDataPacket(net: NetworkManager, packet: SUpdateTileEntityPacket) {
        load(blockState, packet.tag)
    }

    override fun getUpdateTag(): CompoundNBT {
        return save(super.getUpdateTag())
    }

    override fun handleUpdateTag(state: BlockState, tag: CompoundNBT) {
        super.handleUpdateTag(state, tag)
        load(state, tag)
    }

    fun saveAndSync() {
        setChanged()
        level!!.setBlockAndUpdate(worldPosition, blockState)
    }
}